### PR TITLE
test(quantic): Tabs E2E Cypress to Playwright + unit tests Migration

### PR DESCRIPTION
**Pull request by mmitiche**
[SFINT-5794](https://coveord.atlassian.net/browse/SFINT-5794)

### Quantic Tab Bar component:
- Added a `tabBarItem` slot name for the quantic Tab slot
- Added unit tests for the `quanticTabBar` component
- Added Playwright E2E tests for the `quanticTabBar` component

### Quantic Tab component:
- Added unit tests for the `quanticTab` component
- Added Playwright E2E tests for the `quanticTab` component

### UNIT TESTS:

QUANTIC TAB:

image

QUANTIC TAB BAR:

image

### E2E PLAYWRIGHT TESTS:

QUANTIC TAB:

image

QUANTIC TAB BAR:

image

[SFINT-5794]: https://coveord.atlassian.net/browse/SFINT-5794?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by erocheleau** [comment]
@mmitiche do you mind as part of one of the two PRs fix the readme please?

The readme says this:

```
To run Playwright tests, run:

npm run e2e:playwright:
```

There is a `:` at the end that shouldn't be there I think :) I may have missed it in the previous PR.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      const expectedDarkThemeClass = 'slds-theme_shade';
+      const element = createTestComponent({lightTheme: false});
+      await flushPromises();
+
+      const tabBarContainer = element.shadowRoot.querySelector(
+        selectors.tabBarContainer
+      );
+      expect(tabBarContainer.classList).toContain(expectedDarkThemeClass);
+    });
+  });
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:122]: I don't know if you agree but I find we always have too many `describe` xD

```suggestion
  describe('the lightTheme property', () => {
    it('when true, should display the component as light theme', async () => {
      const expectedDarkThemeClass = 'slds-theme_shade';
      const element = createTestComponent({lightTheme: true});
      await flushPromises();

      const tabBarContainer = element.shadowRoot.querySelector(
        selectors.tabBarContainer
      );
      expect(tabBarContainer.classList).not.toContain(expectedDarkThemeClass);
    });

    it('when false, should display the component as dark theme', async () => {
      const expectedDarkThemeClass = 'slds-theme_shade';
      const element = createTestComponent({lightTheme: false});
      await flushPromises();

      const tabBarContainer = element.shadowRoot.querySelector(
        selectors.tabBarContainer
      );
      expect(tabBarContainer.classList).toContain(expectedDarkThemeClass);
    });
  });
```

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  constructor(public page: Page) {
+    this.page = page;
+  }
+
+  get tabBar(): Locator {
+    return this.page.locator('c-quantic-tab-bar');
+  }
+
+  get allVisibleTabs(): Locator {
+    return this.tabBar.locator('button.slds-tabs_default__item').filter({
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:14]: `getByRole` ?

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+const smallViewportWidth = 260;
+const mediumViewportWidth = 300;
+const standardViewportHeight = 1080;
+
+const viewportWidths = [
+  {
+    label: 'extra small',
+    width: extraSmallViewportWidth,
+    primaryExpectation: 'the selected tab ',
+    secondaryExpectation: 'and the more button should be shrunk',
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:19]: Superfluous to create two variables there

```suggestion
    expectationDescription: 'should display only the selected tab, the other tabs should be displayed in the dropdown list and the more button should be shrunk'
```

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    numberOfTabsDisplayed: 1,
+    numberOfOptionsInDropdown: 3,
+    moreButtonLabel: '',
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:22]:
```suggestion
    expectedNumberOfTabsDisplayed: 1,
    expectedNumberOfOptionsInDropdown: 3,
    expectedMoreButtonLabel: '',
```

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  get dropdown(): Locator {
+    return this.page.locator('.slds-dropdown-trigger');
+  }
+
+  get moreButtonLabel(): Locator {
+    return this.moreButton.locator('button').first();
+  }
+
+  get allDropdownOptions(): Locator {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:35]:
```suggestion
  get allTabsInDropdown(): Locator {
```

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            const isMoreDropdownVisible = await tabBar.dropdown.isVisible();
+            expect(isMoreDropdownVisible).toBe(true);
+
+            await tabBar.clickMoreButton();
+
+            const dropdownOptions = tabBar.allDropdownOptions;
+            expect(dropdownOptions).not.toBeNull();
+            expect(await dropdownOptions.count()).toEqual(
+              numberOfOptionsInDropdown
+            );
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:93]: "Options" is fairly generic, I think it's clearer to specify these are tabs.

```suggestion
            const tabsInDropdown = tabBar.allTabsInDropdown;
            expect(tabsInDropdown).not.toBeNull();
            expect(await tabsInDropdown.count()).toEqual(
              numberOfTabsInDropdown
            );
```

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      }) => {
+        const expectedNumberOfDropdownOptions = 2;
+        await tabBar.clickMoreButton();
+
+        const dropdownOptions = tabBar.allDropdownOptions;
+        expect(dropdownOptions).not.toBeNull();
+        expect(await dropdownOptions.count()).toEqual(
+          expectedNumberOfDropdownOptions
+        );
+      });
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:118]: You already tested that no? You already tested a medium width in the previous loop

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+          expect(requestBody).not.toBeNull();
+          expect(requestBody.actionCause).toEqual(expectedActionCause);
+          expect(requestBody.customData.interfaceChangeTo).toEqual(
+            expectedTabValue
+          );
+        });
+      });
+    });
+
+    test.describe('when the dropdown loses focus', () => {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:179]: That really looks like what should be a unit test and not a e2e.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+        const dropdownOptions = tabBar.allDropdownOptions;
+        expect(dropdownOptions).not.toBeNull();
+        expect(await dropdownOptions.count()).toEqual(
+          expectedNumberOfDropdownOptions
+        );
+      });
+
+      test.describe('when a tab is selected from the dropdown list', () => {
+        test('should correctly select the tab and make it active', async ({
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:121]: That's really close to a unit test if it isn't one.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        });
+      }
+    );
+
+    test.describe('when the more button is clicked', () => {
+      // Using medium width (300px), 2 tabs should be displayed and 2 should be in the dropdown.
+      test.use({
+        viewport: {width: mediumViewportWidth, height: standardViewportHeight},
+      });
+      test('should open the dropdown and display the options', async ({
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:107]: This also looks like it should be a unit test.

**Review by erocheleau — CHANGES_REQUESTED** [review]
Some suggested improvements

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+      exampleAssignedElements
+    );
+    await flushPromises();
+
+    exampleSlots.forEach((tabItem, index) => {
+      expect(tabItem.label).toEqual(exampleSlots[index].label);
+      expect(tabItem.engineId).toEqual(exampleSlots[index].engineId);
+      expect(tabItem.expression).toEqual(exampleSlots[index].expression);
+      expect(tabItem.isActive).toEqual(exampleSlots[index].isActive);
+    });
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:92]: what's happening here is checking that `exampleSlots` is equal to `exampleSlots` 🤨

**Comment by mmitiche** [comment]
I added more unit tests for the Quantic Tab Bar component, this component in a bit complicated component, so we should use as much unit tests as possible to test it, we should invest more effort in unit tests in PRs like this.

I have added to this PR the logic to test how the component behaves, by mocking and playing with the width of the container and the slots passed to this component.

File: packages/quantic/decisions/0001-testing-strategy.md [hunk 12]
```diff
@@ -39,12 +39,15 @@ Key points for Playwright E2E tests:
...
+- **Browser-specific behaviors**: Test browser-specific behaviors that would interact with our components.
 
 **Example Playwright Test Scenarios**:
 
 - A user uses a Quantic component by interacting with its multiple buttons and triggering Coveo analytics. We test that the expected analytics are sent, that the events are valid, and the response from the analytics API are correct.
 - A user navigates through a whole search workflow, interacting with multiple components. We test the expected Search API calls and responses, the analytics calls and response, and also the reactivity of the full search experience.
 - Verify the integration between the Quantic components and Salesforce. We test that modifying data through the Salesforce components has an impact and triggers the correct components reactions in a Quantic experience.
+- A user modifies the URL parameters. We assess the impact on our components.
+- The browser viewport size changes. We verify that our components adapt their display accordingly.
```
Comment (mmitiche) [inline packages/quantic/decisions/0001-testing-strategy.md:50]: This point is to be clarified as you can see in this PR, the viewport size logic can be tested with unit tests and sometimes is better tested in unit tests, I see the differentiation as the following: 
- The logic based on the viewport size that makes the component render its element can be be unit tested and is sometimes better unit tested like in this PR.
- The reactivity of the component to the resize events of the browser is something that can be E2E and is better E2E tested.

Bottom line is saying that only E2E tests are used to test such logics is untrue, as I'm noticing while evaluating how the QuanticTabBar component, the unit tests in this PR seem more useful.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/fixture.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+type QuanticTabBarE2EFixtures = {
+  tabBar: TabBarObject;
+  search: SearchObject;
+  options: Partial<TabBarOptions>;
+};
+
+type QuanticTabBarE2ESearchFixtures = QuanticTabBarE2EFixtures & {
+  urlHash: string;
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/fixture.ts:24]: The tab bar should  not need a urlHash

**Comment by mmitiche** [comment]
@SimonMilord  I added unit tests for the QuanticTab component and the QuanticTabBar component, I will let @erocheleau review the rest!

File: packages/quantic/decisions/0001-testing-strategy.md (hunk 12 above)
Comment (erocheleau) [reply packages/quantic/decisions/0001-testing-strategy.md:50]: The VIEWPORT cannot be tested in unit tests, but the size of the container of an element can definitely be tested via unit tests.

That's why I suggested that since this was a bit more related to reactivity (aka how many tabs can be shown in a specific width) we should do those via playwright tests.

But I agree it's in a gray zone.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/quanticTabBar.js [hunk 14]
```diff
@@ -357,7 +357,10 @@ export default class QuanticTabBar extends LightningElement {
    * @returns {Array<Element>}
    */
   getTabsFromSlot() {
-    return Array.from(this.querySelectorAll('*')).filter((element) =>
+    /** @type {HTMLSlotElement} */
+    const slot = this.template.querySelector('slot');
+
+    return Array.from(Array.from(slot.assignedElements())).filter((element) =>
```
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticTabBar/quanticTabBar.js:363]: Why do you even need an Array.from()? `slot.assignedElements()` already return an Element[] it looks like?

**Review by erocheleau — COMMENTED** [review]
Only 2 or 3 questions left I think

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/quanticTabBar.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticTabBar/quanticTabBar.js]: you are right, not sure why it was added, but it works without so Ill remove it

File: packages/quantic/cypress/e2e/default-1/tab-bar/tab-bar-selectors.ts [hunk 15]
```diff
@@ -21,7 +21,8 @@ export const TabBarSelectors: TabBarSelector = {
   allTabs: () => TabBarSelectors.get().find(tabComponent),
   activeTab: () =>
     TabBarSelectors.get().find('button.slds-tabs_default__item.slds-is-active'),
-  moreButton: () => TabBarSelectors.get().find('.tab-bar_more-button'),
+  moreButton: () =>
```
Comment (erocheleau) [inline packages/quantic/cypress/e2e/default-1/tab-bar/tab-bar-selectors.ts:24]: So that one didn't work with the `getByRole` I assume?

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+      const exampleIndex = 0;
+      await tabItemsInDropdown[exampleIndex].click();
+      expect(tabItemsInDropdown.length).toBe(expectedNumberOfTabsToBeHidden);
+      expect(functionMocks.select).toHaveBeenCalledTimes(1);
+      expect(functionMocks.select).toHaveBeenCalledWith(
+        expectedNumberOfTabsToBeVisible + exampleIndex
+      );
+    });
+
+    describe('reactivity to window resize', () => {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:279]: You already have this describe on line 143, I think these tests should be moved in the same describe group.

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  get activeTab(): Locator {
+    return this.tabBar.locator('c-quantic-tab').filter({
+      has: this.page.locator('.slds-is-active'),
+    });
+  }
+
+  get moreButton(): Locator {
+    return this.tabBar.getByTestId('tab-bar_more-button');
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:26]: That didn't work with the `getByRole` right? I think I remember trying to make it work...

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+          await tabBar.clickDropdownOption(1);
+
+          const activeTab = await tabBar.activeTab.textContent();
+
+          expect(activeTab).not.toBeNull();
+          expect(activeTab).toEqual(expectedSelectedTabLabel);
+        });
+
+        test('should trigger a new search send the correct UA analytics event', async ({
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:124]: Merge these two tests:

`test('should make the tab active, trigger a new search and send the correct analytics event')`

You're performing the same "action" in order to test something, so test everything that should happen from the same action.

**Review by erocheleau — APPROVED** [review]
Some small comment and questions, but almost ready to merge!

File: packages/quantic/cypress/e2e/default-1/tab-bar/tab-bar-selectors.ts (hunk 15 above)
Comment (SimonMilord) [reply packages/quantic/cypress/e2e/default-1/tab-bar/tab-bar-selectors.ts:24]: nope, thats why I settled for the `getByTestId`

File: packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts (hunk 17 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:26]: yeah I tried it but it worked with `getByTestId`

**Review by erocheleau — APPROVED** [review]

**Review by erocheleau — APPROVED** [review]
